docs: add comprehensive guide for managing private state and witnesses (#1217)
midnightntwrk/midnight-docs: docs/guides/private-state-witnesses.mdx

---
title: Managing Private State with Witnesses
description: A practical guide to using witnesses for private state management in Compact and the midnight-js SDK.

> sidebar_position: 4

Reviewer: Consider matching the sibling guides’ frontmatter: the SPDX and copyright header, a sentence-case imperative title, a description in the 120 to 160 character range that names the Compact runtime this page actually uses, and tags. `sidebar_position: 4` also collides with the networks guide, and 46 is free next to the JavaScript runtime guide at 45.

```suggestion
SPDX-License-Identifier: Apache-2.0
copyright: This file is part of midnight-docs. Copyright (C) Midnight Foundation. Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
title: Manage private state with witnesses
description: Declare witnesses in Compact, implement them in TypeScript with the Compact runtime, wire up private state, and avoid common witness pitfalls.
sidebar_position: 46
tags: [witnesses, private-state, compact, zero-knowledge, typescript, smart-contracts]
```

---

# Managing Private State with Witnesses

> 

Reviewer: Docusaurus already renders the frontmatter title as the H1, so this heading shows up twice on the published page. Consider removing it.

```suggestion
```

> In the Midnight Network, ensuring data privacy while maintaining verifiable state is achieved through **Zero-Knowledge (ZK) proofs**. A core concept in building Midnight dApps is the **witness**—data that is known to the user (the prover) but kept hidden from the network (the verifiers).

Reviewer: Small style pass: lowercase zero-knowledge mid-sentence, DApps casing, active voice, and a colon instead of the dash. Introducing the witness as a function also matches how Compact declares it.

```suggestion
In the Midnight Network, **zero-knowledge (ZK) proofs** let you keep data private while maintaining verifiable state. A core concept in building Midnight DApps is the **witness**: a function that supplies data known to the user (the prover) but kept hidden from the network (the verifiers).
```

This guide covers how to define and use witnesses in the **Compact DSL**, integrate them using the **Compact JavaScript implementation**, and troubleshoot common issues.

> ---

Reviewer: Consider a prerequisites section before the first code block so readers can follow along, including the runtime version pin the examples on this page match. The horizontal rule can go, the sibling guides rely on headings alone.

```suggestion
This guide covers how to declare witnesses in Compact, implement them in TypeScript against the compiler's generated module, and troubleshoot common issues.

## Prerequisites

- The Compact developer tools installed, with `compact compile` working.
- A Node.js project with `@midnight-ntwrk/compact-runtime@0.16.0` installed. This is the version in the [compatibility matrix](/relnotes/support-matrix), and the examples on this page target its API.
- Familiarity with compiling a contract and importing its generated module. [Using Compact contracts from JavaScript](./compact-javascript-runtime) covers this end to end.
```

> ## Core Concepts

Reviewer: Sentence case for headings, to match the rest of docs/guides.

```suggestion
## Core concepts
```

> Before diving into code, it's important to understand the role of a witness in a Midnight dApp:

Reviewer: DApp is the house spelling, and a lighter opener reads more directly.

```suggestion
Before diving into code, it helps to understand the role of a witness in a Midnight DApp:
```

1. **Public State (Ledger)**: Data stored on the Midnight blockchain that anyone can read.
2. **Private State**: Data kept locally by the user.

> 3. **Witness**: A function or variable that supplies private data to a smart contract circuit during local execution. The circuit uses this data to generate a ZK proof, which is then submitted to the network. The network verifies the proof without ever seeing the witness data.

Reviewer: These three definitions are unordered, so bullets fit better than numbers. Two accuracy tweaks folded in: a witness is a function (not a variable), and it is worth telling readers up front that witness values arrive untrusted and need asserting.

```suggestion
- **Public state (ledger)**: Data stored on the Midnight blockchain that anyone can read.
- **Private state**: Data kept locally by the user's DApp and never sent to the network. Your TypeScript code defines its shape and keeps it up to date.
- **Witness**: A function, declared in Compact and implemented in TypeScript, that supplies private data to a smart contract circuit during local execution. The circuit uses this data to generate a ZK proof, which the DApp submits to the network. The network verifies the proof without ever seeing the witness data.

Private state and witnesses are related but distinct: a witness usually reads from (and may update) your private state, and because a witness value arrives untrusted, the circuit must assert any property of it that the contract relies on.
```

---

> ## Step-by-Step Walkthrough: A Private Access Code

Reviewer: Sentence case here too.

```suggestion
## Step-by-step walkthrough: a private access code
```

> Let's build a simple circuit where a user must provide a secret access code to update a public counter. The network will know the counter increased, but will never know the access code used.

Reviewer: Present tense and a direct opener suit the walkthrough voice.

```suggestion
Build a circuit where a user must provide a secret access code to update a public counter. The network knows the counter increased, but never learns the access code used.
```

> ### 1. Defining the Circuit in Compact

Reviewer: Imperative, sentence-case step headings read better in a numbered walkthrough.

```suggestion
### 1. Define the circuit in Compact
```

> In your Compact contract, you define a `witness` block. This tells the compiler that the circuit expects external, private input during execution.

Reviewer: A witness declaration is a function signature rather than a block, so this avoids coining a construct the language does not have.

```suggestion
In your Compact contract, you declare a `witness` function. This tells the compiler that the circuit expects external, private input during execution, supplied by your TypeScript code at call time.
```

```compact
// access_control.compact

> pragma language_version 0.16;

Reviewer: A bare version in the pragma is an exact-match requirement, so this line only compiles on a 0.16 toolchain, which readers no longer install. I verified `>= 0.16` compiles under both current compilers (0.31.1 and 0.34.0).

```suggestion
pragma language_version >= 0.16;
```

import CompactStandardLibrary;

export ledger counter: Uint<32>;

// The network knows the hash of the secret code, but not the code itself.
// The expected hash is set once at deployment via the constructor.
export ledger secret_hash: Bytes<32>;

> // Define the witness structure

Reviewer: Matching the function terminology from the prose above.

```suggestion
// Declare the witness function
```

witness access_witness(): Bytes<32>;

constructor(sk: Bytes<32>) {
    secret_hash = disclose(persistentHash<Bytes<32>>(sk));
}

export circuit increment_counter(): [] {
    // 1. Fetch the private data via the witness
    const secret_code = access_witness();

    // 2. Verify the secret code matches the expected hash
    // The hash function is evaluated locally, and a proof of correct execution is generated.
    const computed_hash = persistentHash<Bytes<32>>(secret_code);
    assert(computed_hash == secret_hash, "Invalid access code");

    // 3. Update the public ledger

>     counter = counter + 1;

Reviewer: This line does not compile as written: `counter + 1` widens to `Uint<0..4294967297>`, and the compiler rejects the assignment (verified on both current compilers). The cast fixes it.

```suggestion
    counter = (counter + 1) as Uint<32>;
```

}
```

> ### 2. Providing the Witness in TypeScript

Reviewer: Imperative and sentence case, matching step 1.

```suggestion
### 2. Provide the witness in TypeScript
```

When the Compact compiler processes your contract, it generates a JavaScript implementation in the `managed/` directory. The generated `Contract` class requires a **witnesses object** whose keys match the witness names declared in your `.compact` file.

Each witness function receives a `WitnessContext` carrying the current ledger view, private state, and contract address, and returns a tuple of the updated private state and the witness value.

```typescript
import * as RT from '@midnight-ntwrk/compact-runtime';
import { Contract, ledger } from './managed/access_control/contract/index.js';
import type { Ledger } from './managed/access_control/contract/index.js';
import { WitnessContext } from '@midnight-ntwrk/compact-runtime';

// Define the shape of your private state
type AccessControlPrivateState = {
  readonly secretCode: Uint8Array;
};

// Create a factory for initial private state
const createPrivateState = (secretCode: Uint8Array): AccessControlPrivateState => ({
  secretCode,
});

> // Implement witnesses — keys must match Compact witness names exactly

Reviewer: The repo style avoids em-dashes, including in code comments.

```suggestion
// Implement witnesses: keys must match Compact witness names exactly
```

const witnesses = {
  access_witness: ({
    privateState,
  }: WitnessContext<Ledger, AccessControlPrivateState>): [
    AccessControlPrivateState,
    Uint8Array,
  ] => [privateState, privateState.secretCode],
};

// Instantiate the contract with the witnesses object
const contract = new Contract(witnesses);
```

> Once instantiated, build a circuit context with the runtime helpers and call the circuit through `contract.impureCircuits`. Build the context with `RT.createConstructorContext` and `RT.createCircuitContext` rather than by hand — a real `CircuitContext` carries query-context state that the wrappers validate.

Reviewer: Same em-dash rule, and this also varies the doubled "Build ... Build" openers.

```suggestion
Once instantiated, build a circuit context with the runtime helpers and call the circuit through `contract.impureCircuits`. Use `RT.createConstructorContext` and `RT.createCircuitContext` rather than assembling the context by hand: a real `CircuitContext` carries query-context state that the wrappers validate.
```

```typescript
const COIN = '0'.repeat(64);
const ADDR = RT.sampleContractAddress();

> const secretCode = new Uint8Array([/* ... 32 bytes of secret data ... */]);

Reviewer: The placeholder array is zero-length as written, so a copy-paste of this snippet type-checks but throws `expected value of type Bytes<32>` at `initialState` (I ran it). A concrete 32-byte value keeps the example runnable.

```suggestion
const secretCode = new Uint8Array(32).fill(7); // any 32-byte secret
```

// 1. Create the genesis state with the constructor
const ctor = contract.initialState(
  RT.createConstructorContext(createPrivateState(secretCode), COIN),
  secretCode,   // constructor argument: the secret key to hash and store
);

// 2. Build a circuit context from the constructor result

> const ctx = RT.createCircuitContext(

Reviewer: One thing worth deciding here rather than fixing inline: I compiled the contract and ran this exact script, and it works as written against `@midnight-ntwrk/compact-runtime` 0.16.0, the version in the compatibility matrix. On npm latest (0.19.0) the same code no longer compiles: `initialState` and circuit calls became async, `createCircuitContext` takes the circuit name as its first argument, and the state path moves to `call.context.callContext.currentQueryContext.state`. The prerequisites pin I suggested above keeps the page truthful for 0.16.0. If you would rather target the latest runtime instead, happy to share the five corrected lines, they are verified too. Either way, a short note that `sampleContractAddress` and the zeroed coin key are local test scaffolding (not production wiring) would save readers a wrong turn.

  ADDR, COIN, ctor.currentContractState, createPrivateState(secretCode),
);

> // 3. Call the impure circuit — this invokes access_witness() automatically

Reviewer: Same em-dash rule for this comment.

```suggestion
// 3. Call the impure circuit, which invokes access_witness() automatically
```

const call = contract.impureCircuits.increment_counter(ctx);

// 4. Read the resulting ledger state with the generated helper
const state = ledger(call.context.currentQueryContext.state);
console.log('New counter value:', state.counter);
```

For a complete walkthrough of importing the generated module, writing witnesses, calling circuits, and running tests, see [Using Compact contracts from JavaScript](./compact-javascript-runtime).

---

> ## Troubleshooting & Best Practices

Reviewer: Sentence case, and "and" over the ampersand.

```suggestion
## Troubleshooting and best practices
```

When working with witnesses, you might encounter a few common pitfalls. 

> ### 1. "Witness not provided" or "Missing witness" Error

Reviewer: Lowercase "errors" outside the quoted messages.

```suggestion
### 1. "Witness not provided" or "Missing witness" errors
```

**Cause**: The `Contract` constructor validates that every witness declared in Compact has a matching key in the witnesses object. If any key is missing, the constructor throws.
**Solution**: Ensure every key in your witnesses object *exactly* matches the name of the witness defined in your `.compact` file.

> ### 2. Non-Deterministic Witnesses

Reviewer: Sentence case.

```suggestion
### 2. Non-deterministic witnesses
```

**Cause**: If your witness function returns different values if called multiple times during the same execution context, proof generation will fail. The prover evaluates the circuit multiple times to generate constraints.

> **Solution**: Ensure your witness functions are **pure** and **deterministic**. If you need to generate a random value, generate it *once* before calling the contract method, store it in a variable, and have the witness function return that variable.

Reviewer: Consider fixing the doubled "if", describing the failure as a mismatch between local execution and proving, and dropping "pure", since a witness may legitimately return updated private state.

```suggestion
**Cause**: If your witness function returns different values when called more than once for the same circuit call, the values recorded during local execution no longer match the values used during proof generation, and proving fails.
**Solution**: Ensure your witness functions are deterministic for the duration of a call. If you need a random value, generate it *once* before calling the contract method, store it in a variable, and have the witness function return that variable.
```

```typescript
// BAD: Generates a new random value every time it's called
const witnesses = {
  get_random_nonce: ({ privateState }: WitnessContext<Ledger, MyPrivateState>) =>
    [privateState, crypto.getRandomValues(new Uint8Array(32))],
};

// GOOD: Value is fixed for the duration of the execution
const fixedNonce = crypto.getRandomValues(new Uint8Array(32));
const witnesses = {
  get_random_nonce: ({ privateState }: WitnessContext<Ledger, MyPrivateState>) =>
    [privateState, fixedNonce],
};
```

### 3. State Commitment Mismatches
**Cause**: You updated public state in the ledger based on private state, but the private state was invalid according to the circuit's `assert` statements.

> **Solution**: Always validate inputs on the client side before invoking the contract to save time. Proof generation is computationally expensive, so failing early in the UI is better than failing during proof generation.

Reviewer: The cause and solution here describe a failed assert rather than a commitment mismatch, so renaming the item makes the heading match its content.

```suggestion
### 3. Failed asserts during proof generation
**Cause**: The private data your witness supplied did not satisfy the circuit's `assert` statements, for example a secret whose hash does not match the stored `secret_hash`.
**Solution**: Validate inputs on the client side before invoking the contract. Proof generation is computationally expensive, so failing early in the UI is better than failing partway through proving.
```

> ## Next Steps

Reviewer: Sentence case.

```suggestion
## Next steps
```

- Learn more about [Compact data types](../compact/data-types).

> - Explore [Using Compact contracts from JavaScript](./compact-javascript-runtime) for the full generated-module API.

Reviewer: Both links resolve on current main. Two more would round out the journey: the security guide covers trusting witness data, and the deployment guide is the natural step after local circuit calls.

```suggestion
- Learn more about [Compact data types](../compact/data-types).
- Explore [Using Compact contracts from JavaScript](./compact-javascript-runtime) for the full generated-module API.
- Harden contracts that consume witness data with [Security and best practices](./security-best-practices).
- When you are ready for a network, [Deploy and operate a contract](./deploy-and-operate) covers providers, deployment, and maintenance.
```
